share/static/js/chat.js

```diff
@@ -1,65 +1,123 @@
-            $(value).html(sanitizeHTML(current_username));
-          }
-        })
-      }
-      $('#chat').scrollTop($('#chat')[0].scrollHeight);
-    });
-    socket.on('update_list', function(data) {
-      if (username_list.indexOf(data.name) === -1 &&
-        current_username !== data.name &&
-        last_username !== data.name
-      ) {
-        username_list.push(data.name);
-        $('#user-list').append('<li>' + sanitizeHTML(data.name) + '</li>')
-      }
-      $('#chat').scrollTop($('#chat')[0].scrollHeight);
+      addMessageToRoom(data, current_username, 'status');
     });
+
+    // Triggered when message is received from a user. Even when sent
+    // by self, it get triggered after the server sends back the emit.
     socket.on('message', function(data) {
-      $('#chat').append('<p>' + sanitizeHTML(data.msg) + '</p>');
-      $('#chat').scrollTop($('#chat')[0].scrollHeight);
+      addMessageToRoom(data, current_username, 'chat');
     });
+
+    // Trigger new message on enter or click of send message button.
     $('#new-message').on('keypress', function(e) {
       var code = e.keyCode || e.which;
       if (code == 13) {
         emitMessage(socket);
       }
     });
-    $('#send-button').on('click', emitMessage);
+    $('#send-button').on('click', function(e) {
+      emitMessage(socket);
+    });
+
+    // Update username
     $('#update-username').on('click', function() {
       var username = $('#username').val();
+      current_username = username;
       socket.emit('update_username', {username: username});
     });
+
+    // Show warning of losing data
+    $(window).on('beforeunload', function (e) {
+      e.preventDefault();
+      e.returnValue = '';
+      return '';
+    });
   });
 });
 
+var addMessageToRoom = function(data, current_username, messageType) {
+  var scrollDiff = getScrollDiffBefore();
+  if (messageType === 'status') {
+    addStatusMessage(data.msg);
+    if (data.connected_users) {
+      addUserList(data.connected_users, current_username);
+    }
+  } else if (messageType === 'chat') {
+    addChatMessage(data.msg)
+  }
+  scrollBottomMaybe(scrollDiff);
+}
+
 var emitMessage = function(socket) {
   var text = $('#new-message').val();
   $('#new-message').val('');
+  $('#chat').scrollTop($('#chat')[0].scrollHeight);
   socket.emit('text', {msg: text});
 }
 
+/************************************/
+/********* Util Functions ***********/
+/************************************/
+
+var createUserListHTML = function(connected_users, current_user) {
+  var userListHTML = '';
+  connected_users.sort();
+  connected_users.forEach(function(username) {
+    if (username !== current_user) {
+      userListHTML += `<li>${sanitizeHTML(username)}</li>`;
+    }
+  });
+  return userListHTML;
+}
+
+var getScrollDiffBefore = function() {
+  return $('#chat').scrollTop() - ($('#chat')[0].scrollHeight - $('#chat')[0].offsetHeight);
+}
+
+var scrollBottomMaybe = function(scrollDiff) {
```

Comment:
@emmapeel2 this should address the issue that you raised UX wise. Right now the UX logic for scrolling to bottom is something like this:

- If the current user is sending a message, even if they have scrolled up, on sending a message, the chat scrolls to bottom.
- If the current user's chat window is already scrolled to the bottom, any incoming status or chat message will still scroll to the bottom.
- If the current user has scrolled up in the chat window, any incoming status or chat message will not change their scroll position.

I think that's the common norm in chat? Let me know if that sounds good.